tests/utils.cuh

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cuco {
+namespace test {
+namespace detail {
+
+template <typename Iterator, typename Predicate>
+__global__ void count_if(Iterator begin, Iterator end, int* count, Predicate p)
+{
+  auto tid = blockDim.x * blockIdx.x + threadIdx.x;
+  auto it  = begin + tid;
+
+  while (it < end) {
+    atomicAdd(count, static_cast<int>(p(*it)));
+    it += gridDim.x * blockDim.x;
+  }
+}
+
+template <typename Iterator1, typename Iterator2, typename Predicate>
+__global__ void count_if(
+  Iterator1 begin1, Iterator1 end1, Iterator2 begin2, int* count, Predicate p)
+{
+  auto const n = end1 - begin1;
+  auto tid     = blockDim.x * blockIdx.x + threadIdx.x;
+
+  while (tid < n) {
+    auto cmp = begin1 + tid;
+    auto ref = begin2 + tid;
+    atomicAdd(count, static_cast<int>(p(*cmp, *ref)));
+    tid += gridDim.x * blockDim.x;
+  }
+}
+
+template <typename T, typename Operator>
+__global__ void sort(T* data, int const size, Operator op)
```

Comment:
This makes me uneasy. I don't like the idea of having to write and maintain our own sort implementation just for tests.

Reply:
Right, I was reluctant to do this as well. Despite the downside, using our own sort reduces the build time from 40 seconds to 30 seconds. So we stick with `thrust::sort`?

Reply:
What happens if we just use CUB sort directly instead?

Reply:
Using `cub::DeviceRadixSort::SortKeys` is as expensive as `thrust::sort`.

Reply:
Saw this by chance. Wondering if the main issue is CUB/thrust's sort template being instantiated for all `<KeyT, ValueT>` combinations or whether even just a single sort template instantiation was bad enough for you. I would have some thoughts if the former was the issue but I'm afraid that just a single instantiation may be bad enough already.

Reply:
I just tested explicit instantiation of cub sort and didn't see any impact on build time. :(

Reply:
Sorry, I was not very clear. What I actually meant was that **we** are instantiating the sort for multiple types (because we test for multiple types). Basically, what I was wondering: how much would it help us if instead of instantiating the sort for all those types, we would have to instantiate it just once (just assuming). For the sake of measuring compile time, you could just hardcode all sort invocations to the same type, e.g., `sort(reinterpret_cast<uint32_t>(probe_keys.data().get(), ...)`.

Reply:
OK, I see what you mean now. In this case, one single instantiation is already bad enough.